pkg/k8s/resource/resource.go

```diff
@@ -464,44 +468,74 @@ func (r *resource[T]) Events(ctx context.Context, opts ...EventsOpt) <-chan Even
 			}
 			entry := raw.(queueEntry)
 
-			var (
-				// eventDoneSentinel is a heap allocated object referenced by Done().
-				// If Done() is not called, a finalizer set on this object will be invoked
-				// which panics. If Done() is called, the finalizer is unset.
-				eventDoneSentinel = new(bool)
-				event             Event[T]
-			)
-			event.Done = func(err error) {
-				runtime.SetFinalizer(eventDoneSentinel, nil)
-				queue.eventDone(entry, err)
-				r.metricEventProcessed(event.Kind, err == nil)
-			}
-
-			// Add a finalizer to catch forgotten calls to Done().
-			runtime.SetFinalizer(eventDoneSentinel, doneFinalizer)
+			var event Event[T]
 
 			switch entry := entry.(type) {
 			case syncEntry:
 				event.Kind = Sync
-			case deleteEntry:
-				event.Kind = Delete
-				event.Key = entry.key
-				event.Object = entry.obj.(T)
-			case upsertEntry:
+			case upsertOrDeleteEntry:
 				obj, exists, err := store.GetByKey(entry.key)
-				// If the item didn't exist, then it's been deleted and a delete event will
-				// follow soon.
-				if err != nil || !exists {
-					event.Done(nil)
-					continue loop
+				if !exists || err != nil {
+					deletedObj, ok := queue.deletedObjects.Load(entry.key)
+					if ok {
+						if _, ok := knownObjects.Load(entry.key); !ok {
+							// Subscriber had not seen the creation of this object, ignore the deletion.
+							queue.deletedObjects.CompareAndDelete(entry.key, deletedObj)
+							queue.Done(raw)
+							continue loop
+						}
+						event.Kind = Delete
+						event.Key = entry.key
+						event.Object = deletedObj.(T)
+					} else {
+						// Here we hit a race in which the object was removed from store
```

Comment:
We could theoretically have also hit a case where `err != nil` from getting from the store (though I don't think the store can actually error). Nonetheless, if there was a persitent error, I think this would continually requeue the same event

Reply:
Currently the store used here can't error since it's a simple in-memory hash map. I think the idea with having errors here is to be able to use e.g. a store that persists to disk. I'll change this to explicitly handle the unexpected error with a `Fatal`.